### PR TITLE
Create password reset feature.

### PR DESCRIPTION
**Pull request by abhikanojia**
- Password resets controller.
 - add columns to users table.
 - add new method to user_mailer
 - create new and edit views.

### CODE REVIEW

File: app/controllers/password_resets_controller.rb [hunk 1]
```diff
@@ -0,0 +1,52 @@
+class PasswordResetsController < ApplicationController
+  skip_before_action :authorize_user
+  before_action :set_user_by_token, only: [:edit, :update]
+  before_action :validate_token, only: :edit
+
+  def new
+  end
+
+  def create
+    user = User.find_by_email(password_request_params[:email])
```
Comment (sawangupta92) [inline app/controllers/password_resets_controller.rb]: move this in before action

File: app/controllers/password_resets_controller.rb [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    if user
+      user.set_password_reset_token
```
Comment (sawangupta92) [inline app/controllers/password_resets_controller.rb]: move this in above if condition

File: app/controllers/password_resets_controller.rb [hunk 3]
```diff
@@ -0,0 +1,52 @@
+class PasswordResetsController < ApplicationController
+  skip_before_action :authorize_user
+  before_action :set_user_by_token, only: [:edit, :update]
```
Comment (sawangupta92) [inline app/controllers/password_resets_controller.rb]: Also, make sure that a logged in user cannot visit this page

File: app/controllers/password_resets_controller.rb [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+    def set_user_by_token
+      @user = User.find_by(reset_token: params[:token])
+      unless @user
+        redirect_to new_password_reset_path, error: t('.invalid_request')
+      end
+    end
+
+    def password_request_params
+      params.require(:password_reset).permit(:email)
```
Comment (sawangupta92) [inline app/controllers/password_resets_controller.rb]: I don't think that we need to permit this

File: app/models/user.rb [hunk 5]
```diff
@@ -1,6 +1,8 @@
 class User < ApplicationRecord
   EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
   DEFAULT_CREDITS = 5
+  PASSWORD_EXPIRES_MINUTES = 30
+  ONE_MINUTE = 1.minute
```
Comment (sawangupta92) [inline app/models/user.rb]: renam

File: app/models/user.rb [hunk 6]
```diff
@@ -1,6 +1,8 @@
 class User < ApplicationRecord
   EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
   DEFAULT_CREDITS = 5
+  PASSWORD_EXPIRES_MINUTES = 30
```
Comment (sawangupta92) [inline app/models/user.rb]: rename

File: app/models/user.rb [hunk 7]
```diff
@@ -29,6 +32,19 @@ def is_disabled?
     !(self.is_active?) && self.activated_at?
   end
 
+  def set_password_reset_token
+    update(reset_token: SecureRandom.urlsafe_base64, reset_request_sent_at: Time.current)
```
Comment (sawangupta92) [inline app/models/user.rb]: make sure that reset_token is unique

File: app/models/user.rb [hunk 8]
```diff
@@ -29,13 +32,36 @@ def is_disabled?
...
+
   def send_verification_email
     UserMailer.verify_email(id).deliver_now if contributor?
   end
 
   private
 
     def add_account_verification_token
-      self.confirmation_token = SecureRandom.urlsafe_base64
+      self.confirmation_token = loop do
```
Comment (sawangupta92) [inline app/models/user.rb]: also add the uniqueness validation

File: app/views/password_resets/edit.html.erb [hunk 9]
```diff
@@ -0,0 +1,23 @@
+<div class="col-md-6 col-md-offset-3">
+<div class="panel panel-primary">
```
Comment (sawangupta92) [inline app/views/password_resets/edit.html.erb]: indentartion

File: app/controllers/password_resets_controller.rb (hunk 4 above)
Comment (sawangupta92) [reply app/controllers/password_resets_controller.rb]: need to fix this also

File: app/controllers/password_resets_controller.rb [hunk 10]
```diff
@@ -0,0 +1,57 @@
+class PasswordResetsController < ApplicationController
+  # include UserSession
```
[the remark this answers is not in the document]
Comment (abhikanojia) [reply app/controllers/password_resets_controller.rb]: Commented code will be used after mergin to master. will un-comment it after merging.

File: app/controllers/password_resets_controller.rb [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+  before_action :set_user_by_token, only: [:edit, :update]
+  before_action :validate_token, only: :edit
+  # before_action :ensure_no_active_session
+  before_action :set_user_by_email, only: :create
+
+  def new
+  end
+
+  def create
+    @user.set_password_reset_token
```
Comment (sawangupta92) [inline app/controllers/password_resets_controller.rb]: if failed?

**Pull request merged**